src/Tempest/Generation/src/StubFileGenerator.php

```diff
@@ -0,0 +1,150 @@
+        }
+    }
+
+    /**
+     * @param StubFile $stubFile The stub file to use for the generation. It must be of type RAW_FILE.
+     * @param string $targetPath The path where the generated file will be saved including the filename and extension.
+     * @param bool $shouldOverride Whether the generator should override the file if it already exists.
+     * @param array<string, string> $replacements An array of key-value pairs to replace in the stub file.
+     *     The keys are the placeholders in the stub file (e.g. 'dummy-content')
+     *     The values are the replacements for the placeholders (e.g. 'real content')
+     *
+     * @param array<Closure(StringHelper): StringHelper> $manipulations An array of manipulations to apply to the generated file raw content.
+     */
+    public function generateRawFile(
+        StubFile $stubFile,
+        string $targetPath,
+        bool $shouldOverride = false,
+        array $replacements = [],
+        array $manipulations = [],
+    ): void {
+        if ($stubFile->type !== StubFileType::RAW_FILE) {
+            throw new FileGenerationFailedException(sprintf('The stub file must be of type RAW_FILE, "%s" given.', $stubFile->type->name));
+        }
+
+        try {
+            $this->prepareFilesystem($targetPath, $shouldOverride);
+
+            $fileContent = file_get_contents($stubFile->filePath);
+
+            foreach ($replacements as $placeholder => $replacement) {
+                if (! is_string($replacement)) {
+                    continue;
+                }
+
+                $fileContent = str($fileContent)->replace($placeholder, $replacement);
+            }
+
+            // Run all manipulations
+            $fileContent = array_reduce(
+                array: $manipulations,
+                initial: $fileContent,
+                callback: fn (StringHelper $content, Closure $manipulation) => $manipulation($content)
+            );
+
+            file_put_contents($targetPath, $fileContent);
+        } catch (Throwable $throwable) {
+            throw new FileGenerationFailedException(sprintf('The file could not be written. %s', $throwable->getMessage()));
+        }
+    }
+
+    /**
+     * Prepare the directory structure for the new file.
+     * It will delete the target file if it exists and we force the override.
+     *
+     * @param string $targetPath The path where the generated file will be saved including the filename and extension.
+     * @param bool $shouldOverride Whether the generator should override the file if it already exists.
+     *
+     * @throws FileGenerationFailedException If the operation has been aborted.
+     */
+    private function prepareFilesystem(
+        string $targetPath,
+        bool $shouldOverride = false,
+    ): void {
+        // Delete the file if it exists and we force the override
+        if (file_exists($targetPath)) {
```

Comment:
Shouldn't we better do this right before copying the file? 

Let's say something goes wrong in between `prepareFilesystem` and actually copying the file. Then we'll have removed the original file, but haven't been able to copy to new one.

Reply:
I agree

I've changed this